File: 5. Sliding Windows/3. Longest Substring Without Repeating Characters/Solution.cpp
#include <string>
#include <unordered_map>

class Solution {
public:
    int lengthOfLongestSubstring(std::string s) {
        std::unordered_map<char, int> index_map;
        int res = 0;
        int left = 0;
        int right = 0;

        while (right < s.size())
        {
            if (index_map.count(s[right]) != 0 && index_map[s[right]] >= left) left = index_map[s[right]] + 1;
            index_map[s[right]] = right;
            int currLength = right - left + 1;
            if (currLength > res) res = currLength;      
            ++right;
        }

        return res;
    }
};

File: 5. Sliding Windows/3. Longest Substring Without Repeating Characters/Solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res = 0
        index_map = {}
        left = 0
        right = 0

        while right < len(s):

            if s[right] in index_map and index_map[s[right]] >= left:
                left = index_map[s[right]] + 1
            
            index_map[s[right]] = right
            curr_length = right - left + 1
            res = res if res > curr_length else curr_length

            right += 1
        
        return res
